// @zag-js/carousel@0.10.3/dist/carousel.dom.mjs
import { createScope, queryAll } from '@zag-js/dom-query';

const dom = createScope({
  getRootId: (ctx) => ctx.ids?.root ?? `carousel:${ctx.id}`,
  getViewportId: (ctx) => ctx.ids?.viewport ?? `carousel:${ctx.id}:viewport`,
  getSlideId: (ctx, index) => ctx.ids?.slide?.(index) ?? `carousel:${ctx.id}:slide:${index}`,
  getSlideGroupId: (ctx) => ctx.ids?.slideGroup ?? `carousel:${ctx.id}:slide-group`,
  getNextTriggerId: (ctx) => ctx.ids?.nextTrigger ?? `carousel:${ctx.id}:next-trigger`,
  getPrevTriggerId: (ctx) => ctx.ids?.prevTrigger ?? `carousel:${ctx.id}:prev-trigger`,
  getIndicatorGroupId: (ctx) => ctx.ids?.indicatorGroup ?? `carousel:${ctx.id}:indicator-group`,
  getIndicatorId: (ctx, index) => ctx.ids?.indicator?.(index) ?? `carousel:${ctx.id}:indicator:${index}`,
  getRootEl: (ctx) => dom.queryById(ctx, dom.getRootId(ctx)),
  getViewportEl: (ctx) => dom.queryById(ctx, dom.getViewportId(ctx)),
  getSlideGroupEl: (ctx) => dom.queryById(ctx, dom.getSlideGroupId(ctx)),
  getSlideEls: (ctx) => queryAll(dom.getSlideGroupEl(ctx), `[data-part=slide]`)
});

export { dom };
