// @zag-js/date-utils@0.10.3/dist/assertion.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const date = require('@internationalized/date');

function isTodayDate(date$1, timeZone) {
  return date.isToday(date$1, timeZone);
}
function isDateEqual(dateA, dateB) {
  return dateB != null && date.isSameDay(dateA, dateB);
}
function isDateInvalid(date, minValue, maxValue) {
  return minValue != null && date.compare(minValue) < 0 || maxValue != null && date.compare(maxValue) > 0;
}
function isDateDisabled(date, startDate, endDate, minValue, maxValue) {
  return date.compare(startDate) < 0 || date.compare(endDate) > 0 || isDateInvalid(date, minValue, maxValue);
}
function isDateUnavailable(date, isUnavailable, locale, minValue, maxValue) {
  if (!date) {
    return false;
  }
  if (isUnavailable?.(date, locale)) {
    return true;
  }
  return isDateInvalid(date, minValue, maxValue);
}
function isDateOutsideVisibleRange(date, startDate, endDate) {
  return date.compare(startDate) < 0 || date.compare(endDate) > 0;
}
function isPreviousVisibleRangeInvalid(startDate, minValue, maxValue) {
  const prevDate = startDate.subtract({ days: 1 });
  return date.isSameDay(prevDate, startDate) || isDateInvalid(prevDate, minValue, maxValue);
}
function isNextVisibleRangeInvalid(endDate, minValue, maxValue) {
  const nextDate = endDate.add({ days: 1 });
  return date.isSameDay(nextDate, endDate) || isDateInvalid(nextDate, minValue, maxValue);
}

exports.isDateDisabled = isDateDisabled;
exports.isDateEqual = isDateEqual;
exports.isDateInvalid = isDateInvalid;
exports.isDateOutsideVisibleRange = isDateOutsideVisibleRange;
exports.isDateUnavailable = isDateUnavailable;
exports.isNextVisibleRangeInvalid = isNextVisibleRangeInvalid;
exports.isPreviousVisibleRangeInvalid = isPreviousVisibleRangeInvalid;
exports.isTodayDate = isTodayDate;
