// @zag-js/date-picker@0.10.3/dist/date-picker.dom.mjs
import { createScope, query } from '@zag-js/dom-query';

const dom = createScope({
  getGridId: (ctx, id) => ctx.ids?.grid?.(id) ?? `datepicker:${ctx.id}:grid:${id}`,
  getHeaderId: (ctx) => ctx.ids?.header ?? `datepicker:${ctx.id}:header`,
  getContentId: (ctx) => ctx.ids?.content ?? `datepicker:${ctx.id}:content`,
  getCellTriggerId: (ctx, id) => ctx.ids?.cellTrigger?.(id) ?? `datepicker:${ctx.id}:cell-trigger:${id}`,
  getPrevTriggerId: (ctx, view) => ctx.ids?.prevTrigger?.(view) ?? `datepicker:${ctx.id}:prev:${view}`,
  getNextTriggerId: (ctx, view) => ctx.ids?.nextTrigger?.(view) ?? `datepicker:${ctx.id}:next:${view}`,
  getViewTriggerId: (ctx, view) => ctx.ids?.viewTrigger?.(view) ?? `datepicker:${ctx.id}:view:${view}`,
  getClearTriggerId: (ctx) => ctx.ids?.clearTrigger ?? `datepicker:${ctx.id}:clear`,
  getControlId: (ctx) => ctx.ids?.control ?? `datepicker:${ctx.id}:control`,
  getInputId: (ctx) => ctx.ids?.input ?? `datepicker:${ctx.id}:input`,
  getTriggerId: (ctx) => ctx.ids?.trigger ?? `datepicker:${ctx.id}:trigger`,
  getMonthSelectId: (ctx) => ctx.ids?.monthSelect ?? `datepicker:${ctx.id}:month-select`,
  getYearSelectId: (ctx) => ctx.ids?.yearSelect ?? `datepicker:${ctx.id}:year-select`,
  getFocusedCell: (ctx, view = ctx.view) => query(dom.getContentEl(ctx), `[data-part=cell-trigger][data-type=${view}][data-focused]:not([data-outside-range])`),
  getTriggerEl: (ctx) => dom.getById(ctx, dom.getTriggerId(ctx)),
  getContentEl: (ctx) => dom.getById(ctx, dom.getContentId(ctx)),
  getInputEl: (ctx) => dom.getById(ctx, dom.getInputId(ctx)),
  getYearSelectEl: (ctx) => dom.getById(ctx, dom.getYearSelectId(ctx)),
  getMonthSelectEl: (ctx) => dom.getById(ctx, dom.getMonthSelectId(ctx)),
  getClearTriggerEl: (ctx) => dom.getById(ctx, dom.getClearTriggerId(ctx))
});

export { dom };
