// @zag-js/utils@0.10.3/dist/array.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

function toArray(v) {
  if (!v)
    return [];
  return Array.isArray(v) ? v : [v];
}
const fromLength = (length) => Array.from(Array(length).keys());
const first = (v) => v[0];
const last = (v) => v[v.length - 1];
const isEmpty = (v) => v.length === 0;
const has = (v, t) => v.indexOf(t) !== -1;
const add = (v, ...items) => v.concat(items);
const remove = (v, item) => removeAt(v, v.indexOf(item));
const removeAt = (v, i) => {
  if (i > -1)
    v.splice(i, 1);
  return v;
};
function clear(v) {
  while (v.length > 0)
    v.pop();
  return v;
}
function nextIndex(v, idx, opts = {}) {
  const { step = 1, loop = true } = opts;
  const next2 = idx + step;
  const len = v.length;
  const last2 = len - 1;
  if (idx === -1)
    return step > 0 ? 0 : last2;
  if (next2 < 0)
    return loop ? last2 : 0;
  if (next2 >= len)
    return loop ? 0 : idx > len ? len : idx;
  return next2;
}
function next(v, idx, opts = {}) {
  return v[nextIndex(v, idx, opts)];
}
function prevIndex(v, idx, opts = {}) {
  const { step = 1, loop = true } = opts;
  return nextIndex(v, idx, { step: -step, loop });
}
function prev(v, index, opts = {}) {
  return v[prevIndex(v, index, opts)];
}
const chunk = (v, size) => {
  const res = [];
  return v.reduce((rows, value, index) => {
    if (index % size === 0)
      rows.push([value]);
    else
      last(rows)?.push(value);
    return rows;
  }, res);
};

exports.add = add;
exports.chunk = chunk;
exports.clear = clear;
exports.first = first;
exports.fromLength = fromLength;
exports.has = has;
exports.isEmpty = isEmpty;
exports.last = last;
exports.next = next;
exports.nextIndex = nextIndex;
exports.prev = prev;
exports.prevIndex = prevIndex;
exports.remove = remove;
exports.removeAt = removeAt;
exports.toArray = toArray;
