// @zag-js/utils@0.10.3/dist/index.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const array = require('./array.js');
const functions = require('./functions.js');
const guard = require('./guard.js');
const object = require('./object.js');
const warning = require('./warning.js');



exports.add = array.add;
exports.chunk = array.chunk;
exports.clear = array.clear;
exports.first = array.first;
exports.fromLength = array.fromLength;
exports.has = array.has;
exports.isEmpty = array.isEmpty;
exports.last = array.last;
exports.next = array.next;
exports.nextIndex = array.nextIndex;
exports.prev = array.prev;
exports.prevIndex = array.prevIndex;
exports.remove = array.remove;
exports.removeAt = array.removeAt;
exports.toArray = array.toArray;
exports.callAll = functions.callAll;
exports.cast = functions.cast;
exports.noop = functions.noop;
exports.runIfFn = functions.runIfFn;
exports.uuid = functions.uuid;
exports.hasProp = guard.hasProp;
exports.isArray = guard.isArray;
exports.isBoolean = guard.isBoolean;
exports.isDev = guard.isDev;
exports.isFunction = guard.isFunction;
exports.isNumber = guard.isNumber;
exports.isObject = guard.isObject;
exports.isString = guard.isString;
exports.compact = object.compact;
exports.json = object.json;
exports.invariant = warning.invariant;
exports.warn = warning.warn;
