// @zag-js/toast@0.10.3/dist/toast.connect.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const domQuery = require('@zag-js/dom-query');
const toast_anatomy = require('./toast.anatomy.js');
const toast_dom = require('./toast.dom.js');

function connect(state, send, normalize) {
  const isVisible = state.hasTag("visible");
  const isPaused = state.hasTag("paused");
  const pauseOnInteraction = state.context.pauseOnInteraction;
  const placement = state.context.placement;
  return {
    /**
     * The type of the toast.
     */
    type: state.context.type,
    /**
     * The title of the toast.
     */
    title: state.context.title,
    /**
     *  The description of the toast.
     */
    description: state.context.description,
    /**
     * The current placement of the toast.
     */
    placement,
    /**
     * Whether the toast is visible.
     */
    isVisible,
    /**
     * Whether the toast is paused.
     */
    isPaused,
    /**
     * Whether the toast is in RTL mode.
     */
    isRtl: state.context.dir === "rtl",
    /**
     * Function to pause the toast (keeping it visible).
     */
    pause() {
      send("PAUSE");
    },
    /**
     * Function to resume the toast dismissing.
     */
    resume() {
      send("RESUME");
    },
    /**
     * Function to instantly dismiss the toast.
     */
    dismiss() {
      send("DISMISS");
    },
    /**
     * Function render the toast in the DOM (based on the defined `render` property)
     */
    render() {
      return state.context.render?.({
        id: state.context.id,
        type: state.context.type,
        duration: state.context.duration,
        title: state.context.title,
        placement: state.context.placement,
        description: state.context.description,
        dismiss() {
          send("DISMISS");
        }
      });
    },
    rootProps: normalize.element({
      ...toast_anatomy.parts.root.attrs,
      dir: state.context.dir,
      id: toast_dom.dom.getRootId(state.context),
      "data-open": domQuery.dataAttr(isVisible),
      "data-type": state.context.type,
      "data-placement": placement,
      role: "status",
      "aria-atomic": "true",
      tabIndex: 0,
      style: {
        position: "relative",
        pointerEvents: "auto",
        margin: "calc(var(--toast-gutter) / 2)",
        "--remove-delay": `${state.context.removeDelay}ms`,
        "--duration": `${state.context.duration}ms`
      },
      onKeyDown(event) {
        if (event.key == "Escape") {
          send("DISMISS");
          event.preventDefault();
        }
      },
      onFocus() {
        if (pauseOnInteraction) {
          send("PAUSE");
        }
      },
      onBlur() {
        if (pauseOnInteraction) {
          send("RESUME");
        }
      },
      onPointerEnter() {
        if (pauseOnInteraction) {
          send("PAUSE");
        }
      },
      onPointerLeave() {
        if (pauseOnInteraction) {
          send("RESUME");
        }
      }
    }),
    titleProps: normalize.element({
      ...toast_anatomy.parts.title.attrs,
      id: toast_dom.dom.getTitleId(state.context)
    }),
    descriptionProps: normalize.element({
      ...toast_anatomy.parts.description.attrs,
      id: toast_dom.dom.getDescriptionId(state.context)
    }),
    closeTriggerProps: normalize.button({
      id: toast_dom.dom.getCloseTriggerId(state.context),
      ...toast_anatomy.parts.closeTrigger.attrs,
      type: "button",
      "aria-label": "Dismiss notification",
      onClick() {
        send("DISMISS");
      }
    })
  };
}

exports.connect = connect;
