// @zag-js/color-picker@0.10.3/dist/color-picker.dom.mjs
import { getRelativePoint } from '@zag-js/dom-event';
import { createScope, queryAll } from '@zag-js/dom-query';

const dom = createScope({
  getContentId: (ctx) => ctx.ids?.content ?? `color-picker:${ctx.id}:content`,
  getAreaId: (ctx) => ctx.ids?.area ?? `color-picker:${ctx.id}:area`,
  getAreaGradientId: (ctx) => ctx.ids?.areaGradient ?? `color-picker:${ctx.id}:area-gradient`,
  getAreaThumbId: (ctx) => ctx.ids?.areaThumb ?? `color-picker:${ctx.id}:area-thumb`,
  getChannelSliderTrackId: (ctx, channel) => ctx.ids?.channelSliderTrack?.(channel) ?? `color-picker:${ctx.id}:slider-track:${channel}`,
  getChannelInputId: (ctx, channel) => ctx.ids?.channelInput?.(channel) ?? `color-picker:${ctx.id}:input:${channel}`,
  getChannelSliderThumbId: (ctx, channel) => ctx.ids?.channelSliderThumb?.(channel) ?? `color-picker:${ctx.id}:slider-thumb:${channel}`,
  getContentEl: (ctx) => dom.queryById(ctx, dom.getContentId(ctx)),
  getAreaThumbEl: (ctx) => dom.queryById(ctx, dom.getAreaThumbId(ctx)),
  getChannelSliderThumbEl: (ctx, channel) => dom.queryById(ctx, dom.getChannelSliderThumbId(ctx, channel)),
  getChannelInputEl: (ctx, channel) => dom.queryById(ctx, dom.getChannelInputId(ctx, channel)),
  getAreaEl: (ctx) => dom.queryById(ctx, dom.getAreaId(ctx)),
  getAreaValueFromPoint(ctx, point) {
    const { percent } = getRelativePoint(point, dom.getAreaEl(ctx));
    return percent;
  },
  getChannelSliderTrackEl: (ctx, channel) => {
    return dom.queryById(ctx, dom.getChannelSliderTrackId(ctx, channel));
  },
  getChannelSliderValueFromPoint(ctx, point, channel) {
    const { percent } = getRelativePoint(point, dom.getChannelSliderTrackEl(ctx, channel));
    return percent;
  },
  getChannelInputEls: (ctx) => {
    return queryAll(dom.getContentEl(ctx), "input[data-channel]");
  }
});

export { dom };
