// @zag-js/popper@0.10.3/dist/auto-update.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const dom = require('@floating-ui/dom');
const elementRect = require('@zag-js/element-rect');

const callAll = (...fns) => () => fns.forEach((fn) => fn());
const isHTMLElement = (el) => {
  return typeof el === "object" && el !== null && el.nodeType === 1;
};
const addDomEvent = (el, type, fn, options) => {
  el.addEventListener(type, fn, options);
  return () => el.removeEventListener(type, fn, options);
};
function resolveOptions(option) {
  const bool = typeof option === "boolean";
  return {
    ancestorResize: bool ? option : option.ancestorResize ?? true,
    ancestorScroll: bool ? option : option.ancestorScroll ?? true,
    referenceResize: bool ? option : option.referenceResize ?? true
  };
}
function autoUpdate(reference, floating, update, options = false) {
  const { ancestorScroll, ancestorResize, referenceResize } = resolveOptions(options);
  const useAncestors = ancestorScroll || ancestorResize;
  const ancestors = [];
  if (useAncestors && isHTMLElement(reference)) {
    ancestors.push(...dom.getOverflowAncestors(reference));
  }
  function addResizeListeners() {
    let cleanups = [elementRect.trackElementRect(floating, { scope: "size", onChange: update })];
    if (referenceResize && isHTMLElement(reference)) {
      cleanups.push(elementRect.trackElementRect(reference, { onChange: update }));
    }
    cleanups.push(callAll(...ancestors.map((el) => addDomEvent(el, "resize", update))));
    return () => cleanups.forEach((fn) => fn());
  }
  function addScrollListeners() {
    return callAll(...ancestors.map((el) => addDomEvent(el, "scroll", update, { passive: true })));
  }
  return callAll(addResizeListeners(), addScrollListeners());
}

exports.autoUpdate = autoUpdate;
