// @zag-js/popper@0.10.3/dist/get-styles.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const middleware = require('./middleware.js');

const ARROW_FLOATING_STYLE = {
  bottom: "rotate(45deg)",
  left: "rotate(135deg)",
  top: "rotate(225deg)",
  right: "rotate(315deg)"
};
function getPlacementStyles(options) {
  const { placement = "bottom" } = options;
  return {
    arrow: {
      position: "absolute",
      width: middleware.cssVars.arrowSize.reference,
      height: middleware.cssVars.arrowSize.reference,
      [middleware.cssVars.arrowSizeHalf.variable]: `calc(${middleware.cssVars.arrowSize.reference} / 2)`,
      [middleware.cssVars.arrowOffset.variable]: `calc(${middleware.cssVars.arrowSizeHalf.reference} * -1)`
    },
    arrowTip: {
      transform: ARROW_FLOATING_STYLE[placement.split("-")[0]],
      background: middleware.cssVars.arrowBg.reference,
      top: "0",
      left: "0",
      width: "100%",
      height: "100%",
      position: "absolute",
      zIndex: "inherit"
    },
    floating: {
      position: "absolute",
      minWidth: "max-content",
      top: "0px",
      left: "0px"
    }
  };
}

exports.getPlacementStyles = getPlacementStyles;
