// @zag-js/core@0.10.3/dist/create-proxy.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const store = require('@zag-js/store');
const types = require('./types.js');
const functions = require('./packages/utilities/core/src/functions.js');

function createProxy(config) {
  const computedContext = config.computed ?? functions.cast({});
  const initialContext = config.context ?? functions.cast({});
  const state = store.proxy({
    value: config.initial ?? "",
    previousValue: "",
    event: functions.cast({}),
    previousEvent: functions.cast({}),
    context: store.proxyWithComputed(initialContext, computedContext),
    done: false,
    tags: [],
    hasTag(tag) {
      return this.tags.includes(tag);
    },
    matches(...value) {
      return value.includes(this.value);
    },
    can(event) {
      return functions.cast(this).nextEvents.includes(event);
    },
    get nextEvents() {
      const stateEvents = config.states?.[this.value]?.["on"] ?? {};
      const globalEvents = config?.on ?? {};
      return Object.keys({ ...stateEvents, ...globalEvents });
    },
    get changed() {
      if (this.event.value === types.ActionTypes.Init || !this.previousValue)
        return false;
      return this.value !== this.previousValue;
    }
  });
  return functions.cast(state);
}

exports.createProxy = createProxy;
