// @zag-js/avatar@0.10.3/dist/avatar.connect.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const domQuery = require('@zag-js/dom-query');
const avatar_anatomy = require('./avatar.anatomy.js');
const avatar_dom = require('./avatar.dom.js');

function connect(state, send, normalize) {
  const isLoaded = state.matches("loaded");
  const showFallback = !isLoaded;
  return {
    /**
     * Whether the image is loaded.
     */
    isLoaded,
    /**
     * Whether the fallback is shown.
     */
    showFallback,
    /**
     * Function to set new src.
     */
    setSrc(src) {
      send({ type: "SRC.SET", src });
    },
    /**
     * Function to set loaded state.
     */
    setLoaded() {
      send({ type: "IMG.LOADED", src: "api" });
    },
    /**
     * Function to set error state.
     */
    setError() {
      send({ type: "IMG.ERROR", src: "api" });
    },
    rootProps: normalize.element({
      ...avatar_anatomy.parts.root.attrs,
      id: avatar_dom.dom.getRootId(state.context),
      style: {
        display: "grid",
        gridTemplateRows: "1fr 1fr",
        overflow: "hidden"
      }
    }),
    imageProps: normalize.img({
      ...avatar_anatomy.parts.image.attrs,
      id: avatar_dom.dom.getImageId(state.context),
      "data-loaded": domQuery.dataAttr(isLoaded),
      onLoad() {
        send({ type: "IMG.LOADED", src: "element" });
      },
      onError() {
        send({ type: "IMG.ERROR", src: "element" });
      },
      style: {
        gridArea: "1 / 1 / 2 / 2",
        visibility: !isLoaded ? "hidden" : void 0
      }
    }),
    fallbackProps: normalize.element({
      ...avatar_anatomy.parts.fallback.attrs,
      id: avatar_dom.dom.getFallbackId(state.context),
      hidden: !showFallback,
      style: {
        gridArea: "1 / 1 / 2 / 2"
      }
    })
  };
}

exports.connect = connect;
