// @zag-js/slider@0.10.3/dist/slider.machine.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const core = require('@zag-js/core');
const domEvent = require('@zag-js/dom-event');
const domQuery = require('@zag-js/dom-query');
const elementSize = require('@zag-js/element-size');
const formUtils = require('@zag-js/form-utils');
const numericRange = require('@zag-js/numeric-range');
const utils = require('@zag-js/utils');
const slider_dom = require('./slider.dom.js');
const slider_utils = require('./slider.utils.js');

function machine(userContext) {
  const ctx = utils.compact(userContext);
  return core.createMachine(
    {
      id: "slider",
      initial: "idle",
      context: {
        thumbSize: null,
        thumbAlignment: "contain",
        disabled: false,
        threshold: 5,
        dir: "ltr",
        origin: "start",
        orientation: "horizontal",
        initialValue: null,
        value: 0,
        step: 1,
        min: 0,
        max: 100,
        ...ctx
      },
      computed: {
        isHorizontal: (ctx2) => ctx2.orientation === "horizontal",
        isVertical: (ctx2) => ctx2.orientation === "vertical",
        isRtl: (ctx2) => ctx2.orientation === "horizontal" && ctx2.dir === "rtl",
        isInteractive: (ctx2) => !(ctx2.disabled || ctx2.readOnly),
        hasMeasuredThumbSize: (ctx2) => ctx2.thumbSize !== null,
        valuePercent: (ctx2) => 100 * numericRange.getValuePercent(ctx2.value, ctx2.min, ctx2.max)
      },
      watch: {
        value: ["invokeOnChange", "dispatchChangeEvent"]
      },
      activities: ["trackFormControlState", "trackThumbSize"],
      on: {
        SET_VALUE: {
          actions: "setValue"
        },
        INCREMENT: {
          actions: "increment"
        },
        DECREMENT: {
          actions: "decrement"
        }
      },
      entry: ["checkValue"],
      states: {
        idle: {
          on: {
            POINTER_DOWN: {
              target: "dragging",
              actions: ["setPointerValue", "invokeOnChangeStart", "focusThumb"]
            },
            FOCUS: "focus"
          }
        },
        focus: {
          entry: "focusThumb",
          on: {
            POINTER_DOWN: {
              target: "dragging",
              actions: ["setPointerValue", "invokeOnChangeStart", "focusThumb"]
            },
            ARROW_LEFT: {
              guard: "isHorizontal",
              actions: "decrement"
            },
            ARROW_RIGHT: {
              guard: "isHorizontal",
              actions: "increment"
            },
            ARROW_UP: {
              guard: "isVertical",
              actions: "increment"
            },
            ARROW_DOWN: {
              guard: "isVertical",
              actions: "decrement"
            },
            PAGE_UP: {
              actions: "increment"
            },
            PAGE_DOWN: {
              actions: "decrement"
            },
            HOME: {
              actions: "setToMin"
            },
            END: {
              actions: "setToMax"
            },
            BLUR: "idle"
          }
        },
        dragging: {
          entry: "focusThumb",
          activities: "trackPointerMove",
          on: {
            POINTER_UP: {
              target: "focus",
              actions: "invokeOnChangeEnd"
            },
            POINTER_MOVE: {
              actions: "setPointerValue"
            }
          }
        }
      }
    },
    {
      guards: {
        isHorizontal: (ctx2) => ctx2.isHorizontal,
        isVertical: (ctx2) => ctx2.isVertical
      },
      activities: {
        trackFormControlState(ctx2) {
          return formUtils.trackFormControl(slider_dom.dom.getHiddenInputEl(ctx2), {
            onFieldsetDisabled() {
              ctx2.disabled = true;
            },
            onFormReset() {
              if (ctx2.initialValue != null) {
                ctx2.value = ctx2.initialValue;
              }
            }
          });
        },
        trackPointerMove(ctx2, _evt, { send }) {
          return domEvent.trackPointerMove(slider_dom.dom.getDoc(ctx2), {
            onPointerMove(info) {
              send({ type: "POINTER_MOVE", point: info.point });
            },
            onPointerUp() {
              send("POINTER_UP");
            }
          });
        },
        trackThumbSize(ctx2, _evt) {
          if (ctx2.thumbAlignment !== "contain")
            return;
          return elementSize.trackElementSize(slider_dom.dom.getThumbEl(ctx2), (size) => {
            if (size)
              ctx2.thumbSize = size;
          });
        }
      },
      actions: {
        checkValue(ctx2) {
          const value = slider_utils.constrainValue(ctx2, ctx2.value);
          ctx2.value = value;
          ctx2.initialValue = value;
        },
        invokeOnChangeStart(ctx2) {
          ctx2.onChangeStart?.({ value: ctx2.value });
        },
        invokeOnChangeEnd(ctx2) {
          ctx2.onChangeEnd?.({ value: ctx2.value });
        },
        invokeOnChange(ctx2) {
          ctx2.onChange?.({ value: ctx2.value });
        },
        dispatchChangeEvent(ctx2) {
          slider_dom.dom.dispatchChangeEvent(ctx2);
        },
        setPointerValue(ctx2, evt) {
          const value = slider_dom.dom.getValueFromPoint(ctx2, evt.point);
          if (value == null)
            return;
          ctx2.value = numericRange.clampValue(value, ctx2.min, ctx2.max);
        },
        focusThumb(ctx2) {
          domQuery.raf(() => slider_dom.dom.getThumbEl(ctx2)?.focus());
        },
        decrement(ctx2, evt) {
          ctx2.value = slider_utils.decrement(ctx2, evt.step);
        },
        increment(ctx2, evt) {
          ctx2.value = slider_utils.increment(ctx2, evt.step);
        },
        setToMin(ctx2) {
          ctx2.value = ctx2.min;
        },
        setToMax(ctx2) {
          ctx2.value = ctx2.max;
        },
        setValue(ctx2, evt) {
          ctx2.value = slider_utils.constrainValue(ctx2, evt.value);
        }
      }
    }
  );
}

exports.machine = machine;
