// @zag-js/slider@0.10.3/dist/slider.dom.mjs
import { getRelativePoint } from '@zag-js/dom-event';
import { createScope } from '@zag-js/dom-query';
import { dispatchInputValueEvent } from '@zag-js/form-utils';
import { getPercentValue } from '@zag-js/numeric-range';
import { styles } from './slider.style.mjs';

const dom = createScope({
  ...styles,
  getRootId: (ctx) => ctx.ids?.root ?? `slider:${ctx.id}`,
  getThumbId: (ctx) => ctx.ids?.thumb ?? `slider:${ctx.id}:thumb`,
  getControlId: (ctx) => ctx.ids?.control ?? `slider:${ctx.id}:control`,
  getHiddenInputId: (ctx) => ctx.ids?.hiddenInput ?? `slider:${ctx.id}:input`,
  getOutputId: (ctx) => ctx.ids?.output ?? `slider:${ctx.id}:output`,
  getTrackId: (ctx) => ctx.ids?.track ?? `slider:${ctx.id}track`,
  getRangeId: (ctx) => ctx.ids?.track ?? `slider:${ctx.id}:range`,
  getLabelId: (ctx) => ctx.ids?.label ?? `slider:${ctx.id}:label`,
  getMarkerId: (ctx, value) => `slider:${ctx.id}:marker:${value}`,
  getRootEl: (ctx) => dom.getById(ctx, dom.getRootId(ctx)),
  getThumbEl: (ctx) => dom.getById(ctx, dom.getThumbId(ctx)),
  getControlEl: (ctx) => dom.queryById(ctx, dom.getControlId(ctx)),
  getHiddenInputEl: (ctx) => dom.getById(ctx, dom.getHiddenInputId(ctx)),
  getValueFromPoint(ctx, point) {
    const relativePoint = getRelativePoint(point, dom.getControlEl(ctx));
    const percent = relativePoint.getPercentValue({
      orientation: ctx.orientation,
      dir: ctx.dir,
      inverted: { y: true }
    });
    return getPercentValue(percent, ctx.min, ctx.max, ctx.step);
  },
  dispatchChangeEvent(ctx) {
    const input = dom.getHiddenInputEl(ctx);
    if (!input)
      return;
    dispatchInputValueEvent(input, { value: ctx.value });
  }
});

export { dom };
