// @zag-js/checkbox@0.10.3/dist/checkbox.connect.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const domQuery = require('@zag-js/dom-query');
const visuallyHidden = require('@zag-js/visually-hidden');
const checkbox_anatomy = require('./checkbox.anatomy.js');
const checkbox_dom = require('./checkbox.dom.js');

function connect(state, send, normalize) {
  const isDisabled = state.context.disabled;
  const isFocused = !isDisabled && state.context.focused;
  const isChecked = state.context.isChecked;
  const dataAttrs = {
    "data-active": domQuery.dataAttr(state.context.active),
    "data-focus": domQuery.dataAttr(isFocused),
    "data-hover": domQuery.dataAttr(state.context.hovered),
    "data-disabled": domQuery.dataAttr(isDisabled),
    "data-checked": domQuery.dataAttr(state.context.isChecked),
    "data-invalid": domQuery.dataAttr(state.context.invalid),
    "data-indeterminate": domQuery.dataAttr(state.context.isIndeterminate)
  };
  return {
    /**
     * Whether the checkbox is checked
     */
    isChecked,
    /**
     * Whether the checkbox is disabled
     */
    isDisabled,
    /**
     * Whether the checkbox is indeterminate
     */
    isIndeterminate: state.context.isIndeterminate,
    /**
     * Whether the checkbox is focused
     */
    isFocused,
    /**
     *  The checked state of the checkbox
     */
    checkedState: state.context.checked,
    /**
     * Function to set the checked state of the checkbox
     */
    setChecked(checked) {
      send({ type: "DISPATCH.CHANGE", checked });
    },
    /**
     * Function to toggle the checked state of the checkbox
     */
    toggleChecked() {
      send({ type: "CHECKED.TOGGLE", checked: isChecked });
    },
    rootProps: normalize.label({
      ...checkbox_anatomy.parts.root.attrs,
      ...dataAttrs,
      id: checkbox_dom.dom.getRootId(state.context),
      htmlFor: checkbox_dom.dom.getInputId(state.context),
      onPointerMove() {
        if (isDisabled)
          return;
        send({ type: "CONTEXT.SET", context: { hovered: true } });
      },
      onPointerLeave() {
        if (isDisabled)
          return;
        send({ type: "CONTEXT.SET", context: { hovered: false } });
      },
      onPointerDown(event) {
        if (isDisabled)
          return;
        if (isFocused && event.pointerType === "mouse") {
          event.preventDefault();
        }
        send({ type: "CONTEXT.SET", context: { active: true } });
      },
      onPointerUp() {
        if (isDisabled)
          return;
        send({ type: "CONTEXT.SET", context: { active: false } });
      },
      onClick(event) {
        if (event.target === checkbox_dom.dom.getInputEl(state.context)) {
          event.stopPropagation();
        }
      }
    }),
    labelProps: normalize.element({
      ...checkbox_anatomy.parts.label.attrs,
      ...dataAttrs,
      id: checkbox_dom.dom.getLabelId(state.context)
    }),
    controlProps: normalize.element({
      ...checkbox_anatomy.parts.control.attrs,
      ...dataAttrs,
      id: checkbox_dom.dom.getControlId(state.context),
      "aria-hidden": true
    }),
    inputProps: normalize.input({
      ...checkbox_anatomy.parts.input.attrs,
      id: checkbox_dom.dom.getInputId(state.context),
      type: "checkbox",
      required: state.context.required,
      defaultChecked: isChecked,
      disabled: isDisabled,
      "data-disabled": domQuery.dataAttr(isDisabled),
      "aria-labelledby": checkbox_dom.dom.getLabelId(state.context),
      "aria-invalid": state.context.invalid,
      name: state.context.name,
      form: state.context.form,
      value: state.context.value,
      style: visuallyHidden.visuallyHiddenStyle,
      onChange(event) {
        const checked = event.currentTarget.checked;
        send({ type: "CHECKED.SET", checked });
      },
      onBlur() {
        send({ type: "CONTEXT.SET", context: { focused: false } });
      },
      onFocus() {
        send({ type: "CONTEXT.SET", context: { focused: true } });
      },
      onKeyDown(event) {
        if (event.key === " ") {
          send({ type: "CONTEXT.SET", context: { active: true } });
        }
      },
      onKeyUp(event) {
        if (event.key === " ") {
          send({ type: "CONTEXT.SET", context: { active: false } });
        }
      }
    })
  };
}

exports.connect = connect;
