// @zag-js/rect-utils@0.10.3/dist/index.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const align = require('./align.js');
const closest = require('./closest.js');
const contains = require('./contains.js');
const distance = require('./distance.js');
const fromElement = require('./from-element.js');
const fromPoints = require('./from-points.js');
const fromRange = require('./from-range.js');
const fromRotation = require('./from-rotation.js');
const fromWindow = require('./from-window.js');
const getPolygon = require('./get-polygon.js');
const intersection = require('./intersection.js');
const operations = require('./operations.js');
const polygon = require('./polygon.js');
const rect = require('./rect.js');
const union = require('./union.js');



exports.alignRect = align.alignRect;
exports.closest = closest.closest;
exports.closestSideToPoint = closest.closestSideToPoint;
exports.closestSideToRect = closest.closestSideToRect;
exports.contains = contains.contains;
exports.containsPoint = contains.containsPoint;
exports.containsRect = contains.containsRect;
exports.distance = distance.distance;
exports.distanceBtwEdges = distance.distanceBtwEdges;
exports.distanceFromPoint = distance.distanceFromPoint;
exports.distanceFromRect = distance.distanceFromRect;
exports.getElementRect = fromElement.getElementRect;
exports.getRectFromPoints = fromPoints.getRectFromPoints;
exports.fromRange = fromRange.fromRange;
exports.getRotationRect = fromRotation.getRotationRect;
exports.rotate = fromRotation.rotate;
exports.toRad = fromRotation.toRad;
exports.getViewportRect = fromWindow.getViewportRect;
exports.getWindowRect = fromWindow.getWindowRect;
exports.getElementPolygon = getPolygon.getElementPolygon;
exports.collisions = intersection.collisions;
exports.intersection = intersection.intersection;
exports.intersects = intersection.intersects;
exports.expand = operations.expand;
exports.inset = operations.inset;
exports.isSymmetric = operations.isSymmetric;
exports.shift = operations.shift;
exports.shrink = operations.shrink;
exports.debugPolygon = polygon.debugPolygon;
exports.isPointInPolygon = polygon.isPointInPolygon;
exports.createRect = rect.createRect;
exports.getRectCenters = rect.getRectCenters;
exports.getRectCorners = rect.getRectCorners;
exports.getRectEdges = rect.getRectEdges;
exports.isRect = rect.isRect;
exports.union = union.union;
