// @zag-js/dom-event@0.10.3/dist/assertion.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

function isKeyboardClick(e) {
  return e.detail === 0 || e.clientX === 0 && e.clientY === 0;
}
function isPrintableKey(e) {
  return e.key.length === 1 && !e.ctrlKey && !e.metaKey;
}
function isVirtualPointerEvent(e) {
  return e.width === 0 && e.height === 0 || e.width === 1 && e.height === 1 && e.pressure === 0 && e.detail === 0 && e.pointerType === "mouse";
}
function isVirtualClick(e) {
  if (e.mozInputSource === 0 && e.isTrusted)
    return true;
  return e.detail === 0 && !e.pointerType;
}
const isLeftClick = (e) => e.button === 0;
const isContextMenuEvent = (e) => {
  return e.button === 2 || isCtrlKey(e) && e.button === 0;
};
const isModifiedEvent = (e) => e.ctrlKey || e.altKey || e.metaKey;
const isMac = () => /Mac|iPod|iPhone|iPad/.test(window.navigator.platform);
const isCtrlKey = (e) => isMac() ? e.metaKey && !e.ctrlKey : e.ctrlKey && !e.metaKey;

exports.isContextMenuEvent = isContextMenuEvent;
exports.isCtrlKey = isCtrlKey;
exports.isKeyboardClick = isKeyboardClick;
exports.isLeftClick = isLeftClick;
exports.isModifiedEvent = isModifiedEvent;
exports.isPrintableKey = isPrintableKey;
exports.isVirtualClick = isVirtualClick;
exports.isVirtualPointerEvent = isVirtualPointerEvent;
