// @zag-js/dom-event@0.10.3/dist/index.js
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const addDomEvent = require('./add-dom-event.js');
const assertion = require('./assertion.js');
const fireEvent = require('./fire-event.js');
const getEventKey = require('./get-event-key.js');
const getEventPoint = require('./get-event-point.js');
const getEventStep = require('./get-event-step.js');
const getNativeEvent = require('./get-native-event.js');
const getPointValue = require('./get-point-value.js');
const requestPointerLock = require('./request-pointer-lock.js');
const trackPointerMove = require('./track-pointer-move.js');
const trackVisualViewport = require('./track-visual-viewport.js');



exports.addDomEvent = addDomEvent.addDomEvent;
exports.isContextMenuEvent = assertion.isContextMenuEvent;
exports.isCtrlKey = assertion.isCtrlKey;
exports.isKeyboardClick = assertion.isKeyboardClick;
exports.isLeftClick = assertion.isLeftClick;
exports.isModifiedEvent = assertion.isModifiedEvent;
exports.isPrintableKey = assertion.isPrintableKey;
exports.isVirtualClick = assertion.isVirtualClick;
exports.isVirtualPointerEvent = assertion.isVirtualPointerEvent;
exports.fireBlurEvent = fireEvent.fireBlurEvent;
exports.fireCustomEvent = fireEvent.fireCustomEvent;
exports.getEventKey = getEventKey.getEventKey;
exports.getEventPoint = getEventPoint.getEventPoint;
exports.getEventStep = getEventStep.getEventStep;
exports.getNativeEvent = getNativeEvent.getNativeEvent;
exports.getRelativePoint = getPointValue.getRelativePoint;
exports.requestPointerLock = requestPointerLock.requestPointerLock;
exports.trackPointerMove = trackPointerMove.trackPointerMove;
exports.trackVisualViewport = trackVisualViewport.trackVisualViewport;
